cpp/tests/utilities/column_wrapper.hpp

```diff
@@ -944,18 +921,81 @@ class lists_column_wrapper : public detail::column_wrapper {
                  "Unexpected type");
 
     std::vector<size_type> offsetv;
-    offsetv.push_back(0);
-    if (c->size() > 0) { offsetv.push_back(c->size()); }
+    if (c->size() > 0) {
+      offsetv.push_back(0);
+      offsetv.push_back(c->size());
+    }
     auto offsets =
       cudf::test::fixed_width_column_wrapper<size_type>(offsetv.begin(), offsetv.end()).release();
 
     // construct the list column. mark this as a root
-    root    = true;
-    wrapped = make_lists_column(
-      offsetv.size() - 1, std::move(offsets), std::move(c), 0, rmm::device_buffer{0});
+    root  = true;
+    depth = 0;
+
+    size_type num_elements = offsets->size() == 0 ? 0 : offsets->size() - 1;
+    wrapped =
+      make_lists_column(num_elements, std::move(offsets), std::move(c), 0, rmm::device_buffer{0});
+  }
+
+  std::unique_ptr<column> normalize_column(column_view const& col,
+                                           column_view const& expected_hierarchy)
+  {
+    // if are at the bottom of the short column, it must be empty
+    if (col.type().id() != type_id::LIST) {
+      CUDF_EXPECTS(col.size() == 0, "Encountered mismatched column!");
+
+      auto remainder = empty_like(expected_hierarchy);
+      return std::move(remainder);
+    }
+
+    lists_column_view lcv(col);
+    return make_lists_column(col.size(),
+                             std::make_unique<column>(lcv.offsets()),
+                             normalize_column(lists_column_view(col).child(),
+                                              lists_column_view(expected_hierarchy).child()),
+                             UNKNOWN_NULL_COUNT,
+                             copy_bitmask(col));
   }
 
+  std::pair<std::vector<column_view>, std::vector<std::unique_ptr<column>>> preprocess_columns(
+    std::initializer_list<lists_column_wrapper<T>> const& elements,
+    column_view& expected_hierarchy,
+    int expected_depth)
+  {
+    std::vector<std::unique_ptr<column>> stubs;
+    std::vector<column_view> cols;
+
+    // preprocess the incoming lists.
+    // - unwrap any "root" lists
+    // - handle incomplete hierarchies
+    std::transform(elements.begin(),
+                   elements.end(),
+                   std::back_inserter(cols),
+                   [&](lists_column_wrapper const& l) -> column_view {
+                     // depth mismatch.  attempt to normalize the short column.
+                     // this function will also catch if this is a legitmately broken
+                     // set of input
+                     if (l.depth < expected_depth) {
+                       if (l.root) {
+                         CUDF_EXPECTS(l.wrapped->size() == 0, "Mismatch in column types!");
```

Comment:
This message does not appear to match the expects expression.

Reply:
It's unintuitive but correct.  Basically, this is detecting the case where the user has passed in mismatched columns with > 0 values at the level that is mismatched.  Specifically, this is ok:

`{ {{{1, 2, 3}}},  {}  }`
The mismatch here is at the first level and is ok, because the mismatched row (index 1) is empty and can be interpreted as a List<List<List<int>>> that matches row 0.

`{ {{{1, 2, 3}}},  {4, 5, 6}  }`
But this is not ok because it implies the user is passing a concrete List<int> for row 1 instead of the expected List<List<List<int>>> we see from row 0.

Reply:
That's kinda cool. Perhaps a nice comment then?

Reply:
Done.